### PR TITLE
DOC Document the PEFT checkpoint format

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Document the PEFT checkpoint format, with a focus on how to convert to this format.

[Link to rendered docs](https://moon-ci-docs.huggingface.co/docs/peft/pr_1717/en/developer_guides/checkpoint#checkpoint-format).

This is a bit of a brain dump, so let me know if there is more to add or to remove.

### CODE REVIEW

File: docs/source/developer_guides/checkpoint.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+specific language governing permissions and limitations under the License.
+
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Checkpoint format
+
+This document describes the format of the save files used by 🤗 PEFT. Read this if you're interested in how the checkpoint files are structured and how to possibly convert between the PEFT format and other formats.
```
Comment (stas00) [inline docs/source/developer_guides/checkpoint.md:19]:
```suggestion
This document describes the format of the saved files used by 🤗 PEFT. Read this if you're interested in how the checkpoint files are structured and how to possibly convert between the PEFT format and other formats.
```

File: docs/source/developer_guides/checkpoint.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+1. `adapter_model.safetensors` or `adapter_model.bin`
+2. `adapter_config.json`
+3. `README.md`
+
+By default, the model will be saved in the `safetensors` format, which has several advantages over the `bin` format (which uses `pickle` under the hood). Content-wise, the two store the same `state_dict` though and are thus interchangeable.
+
+As mentioned above, the `state_dict` will only contain the parameters of the adapter module, not the base model. To illustrate the difference in size, a normal BERT model requires ~420MB of disk space, whereas an IA³ adapter on top of this BERT model only requires ~260KB.
+
+The `adapter_config.json` file contains the configuration of the adapter module, which is necessary to load the model. Below is an example of a an `adapter_config.json` for an IA³ adapter with standard settings applied to a BERT model:
```
Comment (stas00) [inline docs/source/developer_guides/checkpoint.md:35]:
```suggestion
The `adapter_config.json` file contains the configuration of the adapter module, which is necessary to load the model. Below is an example of an `adapter_config.json` for an IA³ adapter with standard settings applied to a BERT model:
```

File: docs/source/developer_guides/checkpoint.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        # Mark the weight as unmerged
+        self._disable_adapters = False
+        self.merged_adapters = []
+        self.use_dora: dict[str, bool] = {}
+        self.lora_magnitude_vector: Optional[torch.nn.ParameterDict] = None  # for DoRA
+        self._caches: dict[str, Any] = {}
+        self.kwargs = kwargs
+```
+
+Here we can see part of the `__init__` code used by all `LoraLayer` classes in PEFT. There are a bunch of parameters used the initialize the model, but only a few are relevant for the checkpoint file, namely: `lora_A`, `lora_B`, `lora_embedding_A`, and `lora_embedding_B`. These parameters are listed in the class attribute `adapter_layer_names` and they contain the learnable parameters, hence the parameters we need to include in the checkpoint file. All the other parameters, e.g. the rank `r`, are derived from the `adapter_config.json` and have to be included there (unless the default value is used). When converting a LoRA checkpoint from another format to PEFT, it is thus necessary to include those parameters.
```
Comment (stas00) [inline docs/source/developer_guides/checkpoint.md:111]:
```suggestion
Here we can see part of the `__init__` code used by all `LoraLayer` classes in PEFT. There is a bunch of parameters used the initialize the model, but only a few are relevant for the checkpoint file, namely: `lora_A`, `lora_B`, `lora_embedding_A`, and `lora_embedding_B`. These parameters are listed in the class attribute `adapter_layer_names` and they contain the learnable parameters, hence the parameters we need to include in the checkpoint file. All the other parameters, e.g. the rank `r`, are derived from the `adapter_config.json` and have to be included there (unless the default value is used). When converting a LoRA checkpoint from another format to PEFT, it is thus necessary to include those parameters.
```

**Review by stas00 — APPROVED** [review]
Fantastic work, @BenjaminBossan - thank you!

I proposed a few typo fixes

File: docs/source/developer_guides/checkpoint.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+PEFT is all about parameter efficient fine-tuning. This means that generally, PEFT methods will not update each of the model's parameters but only a small subset. This has the nice advantage that the checkpoint files can generally be much smaller than the original model files and are thus easier to store and share. However, this also means that to load a PEFT model, you need to have the original model available as well.
+
+When you call `save_pretrained` on a PEFT model, the PEFT model will save the following files:
+
+1. `adapter_model.safetensors` or `adapter_model.bin`
+2. `adapter_config.json`
+3. `README.md`
+
+By default, the model will be saved in the `safetensors` format, which has several advantages over the `bin` format (which uses `pickle` under the hood). Content-wise, the two store the same `state_dict` though and are thus interchangeable.
```
Comment (younesbelkada) [inline docs/source/developer_guides/checkpoint.md:31]:
```suggestion
By default, the model will be saved in the `safetensors` format, which has several advantages over the `bin` format (which uses `pickle` under the hood). Content-wise, the two store the same `state_dict` though and are thus interchangeable. You can read more about `pickle` file vulnerabilities [in this documentation page.](https://huggingface.co/docs/hub/security-pickle)
```
I attached a nice doc page about pickle vulnerabiltiies

**Review by younesbelkada — APPROVED** [review]
Fantastic work @BenjaminBossan ! Thanks for shipping this !

File: docs/source/developer_guides/checkpoint.md (hunk 4 above)
Comment (BenjaminBossan) [reply docs/source/developer_guides/checkpoint.md]: Thanks.

File: docs/source/developer_guides/checkpoint.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+When we call `save_pretrained`, we strip this adapter name from the keys. The reason for this is that the adapter name is not a fundamental part of the model architecture but just an arbitrary name. When loading the adapter, we could choose a totally different name, and the model would still work the same way. Therefore, the adapter name is not stored in the checkpoint file.
+
+<Tip>
+
+If we call `save_pretrained("some/path")` and the adapter name is not `"default"`, we store the adapter in a sub-directory with the same name as the adapter. So if the name is "other", it would be stored inside of `some/path/other`.
+
+</Tip>
+
+In some circumstances, deciding which values to add to the checkpoint file can become a bit more complicated. For example, in PEFT, DoRA is implemented as a special case of LoRA. Therefore, if you want to convert a DoRA model to PEFT, you should create a LoRA checkpoint with extra entries for DoRA. Referring back to the code above, we can see the following line:
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:146]:
```suggestion
In some circumstances, deciding which values to add to the checkpoint file can become a bit more complicated. For example, in PEFT, DoRA is implemented as a special case of LoRA. If you want to convert a DoRA model to PEFT, you should create a LoRA checkpoint with extra entries for DoRA. You can see this in the `__init__` of the previous `LoraLayer` code:
```

File: docs/source/developer_guides/checkpoint.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+```python
+        self.lora_magnitude_vector: Optional[torch.nn.ParameterDict] = None  # for DoRA
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:149]:
```suggestion
self.lora_magnitude_vector: Optional[torch.nn.ParameterDict] = None  # for DoRA
```

File: docs/source/developer_guides/checkpoint.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+```
+
+This indicates that we can have an optional extra parameter per layer for DoRA.
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:152]:
```suggestion
This indicates that there is an optional extra parameter per layer for DoRA.
```

File: docs/source/developer_guides/checkpoint.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+### Creating the `adapter_config.json`
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:154]:
```suggestion
### adapter_config
```

File: docs/source/developer_guides/checkpoint.md [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+All the other information we need to load  PEFT model is contained in the `adapter_config.json` file. Let's check this file for a LoRA model applied to BERT:
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:156]:
```suggestion
All the other information needed to load a PEFT model is contained in the `adapter_config.json` file. Let's check this file for a LoRA model applied to BERT:
```

File: docs/source/developer_guides/checkpoint.md [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    "query",
+    "value"
+  ],
+  "task_type": null,
+  "use_dora": false,
+  "use_rslora": false
+}
+```
+
+This contains a lot of entries and at first glance, it could feel difficult to figure out all the right values to put in there. However, most of the entries are not necessary to load the model. This is because they use the default values and thus don't need to be added, or because they only affect the initialization of the LoRA weights, which is irrelevant when it comes to loading the model. If you find that you don't know what a specific parameter does, e.g. `"use_rslora"`, just don't add it and you should be fine. Also note that as we add more options, this file will get more entries in the future, but it should be backwards compatible.
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:193]:
```suggestion
This contains a lot of entries, and at first glance, it could feel overwhelming to figure out all the right values to put in there. However, most of the entries are not necessary to load the model. This is either because they use the default values and don't need to be added or because they only affect the initialization of the LoRA weights, which is irrelevant when it comes to loading the model. If you find that you don't know what a specific parameter does, e.g., `"use_rslora",` don't add it, and you should be fine. Also note that as more options are added, this file will get more entries in the future, but it should be backward compatible.
```

File: docs/source/developer_guides/checkpoint.md [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+
+When it comes to the bare minimum of what needs to be included, it is sufficient to have this minimal json:
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:195]:
```suggestion
At the minimum, you should include the following entries:
```

File: docs/source/developer_guides/checkpoint.md [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+
+```json
+{
+  "target_modules": ["query", "value"],
+  "peft_type": "LORA"
+}
+```
+
+However, it is recommended to add as many entries as you can, like the rank `r` or the `base_model_name_or_path` if it's a transformers model. This information can help to understand the model better and to share it more easily with others. To check which keys and values are expected, check out the corresponding `config.py` file in the PEFT source code. For example, for LoRA, this can be found in [`peft/tuners/lora/config.py`](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/config.py).
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:204]:
```suggestion
However, adding as many entries as possible, like the rank `r` or the `base_model_name_or_path` (if it's a Transformers model) is recommended. This information can help others understand the model better and share it more easily. To check which keys and values are expected, check out the [config.py](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/config.py) file (as an example, this is the config file for LoRA) in the PEFT source code.
```

File: docs/source/developer_guides/checkpoint.md [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+
+## Storing the whole PEFT model including the base weights
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:206]:
```suggestion
## Model storage
```

File: docs/source/developer_guides/checkpoint.md [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+
+In some circumstances, you might want to store the whole PEFT model including the base weights. This can be necessary if, for instance, the base model is not available to the users trying to load the PEFT model.
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:208]:
```suggestion
In some circumstances, you might want to store the whole PEFT model, including the base weights. This can be necessary if, for instance, the base model is not available to the users trying to load the PEFT model. You can merge the weights first or convert it into a Transformer model.
```

File: docs/source/developer_guides/checkpoint.md [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+
+### Storing the whole model by merging the weights first
```
Comment (stevhliu) [inline docs/source/developer_guides/checkpoint.md:210]:
```suggestion
### Merge the weights
```

**Review by stevhliu — APPROVED** [review]
Great work as usual!

My main feedback is mostly about centering the doc on converting *to PEFT format* rather than *to and from other formats*, and making things a bit less wordy where appropriate 🙂
